### PR TITLE
[3.0] destroy executor and improve tests 0926

### PR DESCRIPTION
**Pull request by kylixs**
## What is the purpose of the change

* Destroy executor services
* `dubbo-config-spring`  surefire plugin  `reuseForks=true`
* Fix unregister `DubboShutdownHook` error
* Fix `DecodeableRpcResult`  read config NPE
* Improve registry center of tests: PropertySourcesInJavaConfigTest, JavaConfigReferenceBeanTest

### CODE REVIEW

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/RouterChain.java [hunk 1]
```diff
@@ -271,14 +272,23 @@ public void loop(boolean notify) {
         if (firstBuildCache.compareAndSet(true,false)) {
             buildCache(notify);
         }
-        if (notify) {
-            if (loopPermitNotify.tryAcquire()) {
-                loopPool.submit(new NotifyLoopRunnable(true, loopPermitNotify));
+
+        try {
+            if (notify) {
```
Comment (DiegoKrupitza) [inline dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/RouterChain.java:277]: Nested `if`s make the code unreadable. Here would be better to merge the nested `if`s into a single `if`.
The same applies to the `if` in the `else` statement.

I would suggest:
```
if( notify && loopPermitNotify.tryAcquire() ) {
  loopPool.submit(new NotifyLoopRunnable(true, loopPermitNotify));
}else if(loopPermit.tryAcquire()) {
  loopPool.submit(new NotifyLoopRunnable(false, loopPermit));
}
```

**Review by DiegoKrupitza — COMMENTED** [review]
The rest LGTM

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/RouterChain.java (hunk 1 above)
Comment (kylixs) [reply dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/RouterChain.java:277]: The above code logic is not exactly the same as the original one:
if notify==true and loopPermitNotify.tryAcquire() == false,  **no attempt to acquire the loopPermit lock, and no operation will be performed.**

Comment (DiegoKrupitza) [reply dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/RouterChain.java:277]: My suggestion:
```
if( notify && loopPermitNotify.tryAcquire() ) {
  loopPool.submit(new NotifyLoopRunnable(true, loopPermitNotify));
}else if(loopPermit.tryAcquire()) {
  loopPool.submit(new NotifyLoopRunnable(false, loopPermit));
}
```

Current code:
```
if (notify) {
                if (loopPermitNotify.tryAcquire()) {
                    loopPool.submit(new NotifyLoopRunnable(true, loopPermitNotify));
                }
            } else {
                if (loopPermit.tryAcquire()) {
                    loopPool.submit(new NotifyLoopRunnable(false, loopPermit));
                }
            }
```

Semantically they are the same. See the following table with all possible outcomes of that code.

| notify | loopPermitNotify.tryAcquire() | loopPermit.tryAcquire() | Current Code                           | My Suggestion                              | is same |
|--------|-------------------------------|-------------------------|-----------------------------------------------|--------------------------------------------|---------|
| true   | true                          | true                    | NotifyLoopRunnable ( true , loopPermitNotify) | NotifyLoopRunnable(true, loopPermitNotify) | ✅       |
| false  | true                          | true                    | NotifyLoopRunnable ( false , loopPermit)      | NotifyLoopRunnable(false, loopPermit)      | ✅       |
| true   | false                         | true                    | NotifyLoopRunnable ( false , loopPermit)      | NotifyLoopRunnable(false, loopPermit)      | ✅       |
| false  | false                         | true                    | NotifyLoopRunnable ( false , loopPermit)      | NotifyLoopRunnable(false, loopPermit)      | ✅       |
| true   | true                          | false                   | NotifyLoopRunnable ( true , loopPermitNotify) | NotifyLoopRunnable(true, loopPermitNotify) | ✅       |
| false  | true                          | false                   | NO ACTION                                     | NO ACTION                                  | ✅       |
| true   | false                         | false                   | NO ACTION                                     | NO ACTION                                  | ✅       |
| false  | false                         | false                   | NO ACTION                                     | NO ACTION                                  | ✅       |

**Pull request merged**